python-api/livy/client.py

```diff
@@ -0,0 +1,444 @@
+        jobs to get completed
+
+        Parameters
+        ----------
+        shutdown_context : Boolean
+            Whether to shutdown the underlying Spark context. If false, the
+            context will keep running and it's still possible to send commands
+            to it, if the backend being used supports it.
+        """
+        with self.lock:
+            if not self._stopped:
+                self._executor.shutdown(wait=False)
+                try:
+                    if shutdown_context:
+                        session_uri = "/" + str(self._session_id)
+                        headers = {'X-Requested-By': 'livy'}
+                        self._conn.send_request("DELETE", session_uri,
+                                                headers=headers)
+                except:
+                    raise Exception(traceback.format_exc())
+                self._stopped = True
+
+    def _set_uri(self, uri):
+        if uri is not None and uri.scheme in ('http', 'https'):
+            self._config.set(self._CONFIG_SECTION, 'livy.uri', uri.geturl())
+        else:
+            url_exception = uri.geturl if uri is not None else None
+            raise ValueError('Cannot create client - Uri not supported - ',
+                             url_exception)
+
+    def _set_conf(self, key, value):
+        if value is not None:
+            self._config.set(self._CONFIG_SECTION, key, value)
+        else:
+            self._delete_conf(key)
+
+    def _delete_conf(self, key):
+        self._config.remove_option(self._CONFIG_SECTION, key)
+
+    def _set_multiple_conf(self, conf_dict):
+        for key, value in conf_dict.iteritems():
+            self._set_conf(key, value)
+
+    def _load_config(self, load_defaults=True, conf_dict=None):
+        self._config.add_section(self._CONFIG_SECTION)
+        if load_defaults:
+            self._load_default_config()
+        if conf_dict is not None and len(conf_dict) > 0:
+            self._set_multiple_conf(conf_dict)
+
+    def _load_default_config(self):
+        config_dir = os.environ.get(self._LIVY_CLIENT_CONF_DIR)
+        if config_dir is None:
+            raise KeyError('Config directory not set in environment')
+        config_files = os.listdir(config_dir)
+        default_conf_files = ['spark-defaults.conf', 'livy-client.conf']
+        for default_conf_file in default_conf_files:
+            if default_conf_file in config_files:
+                self._load_config_from_files(config_dir, default_conf_file)
+
+    def _load_config_from_files(self, config_dir, config_file):
+        path = os.path.join(config_dir, config_file)
+        data = "[" + self._CONFIG_SECTION + "]\n" + open(path).read()
+        self._config.readfp(StringIO(data.decode('utf8')))
+
+    def _create_new_session(self, session_conf_dict):
+        json_data = json.dumps({'kind': 'pyspark', 'conf': session_conf_dict})
+        header = {'Content-Type': 'application/json', 'X-Requested-By': 'livy'}
```

Comment:
I think it would be better to have a `_send_request` method in this class that had this code to initialize the needed headers. Would avoid a lot of copy & paste.